### PR TITLE
refactor LLM code to langchain runnables and chains

### PR DESCRIPTION
**Pull request by snopoke**

### CODE REVIEW

File: apps/service_providers/llm_service/callbacks.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+        super().__init__()
+        self._lock = threading.Lock()
+        self.model = model
+
+    def __repr__(self) -> str:
+        return f"Prompt Tokens: {self.prompt_tokens}\n" f"Completion Tokens: {self.completion_tokens}\n"
+
+    def on_llm_start(self, serialized: Dict[str, Any], prompts: List[str], **kwargs) -> Any:
+        with self._lock:
+            self.prompt_tokens += sum([self.model.get_num_tokens(prompt) for prompt in prompts])
```
Comment (proteusvacuum) [inline apps/service_providers/llm_service/callbacks.py:25]: nice 👍🏽

File: apps/service_providers/llm_service/main.py [hunk 2]
```diff
@@ -16,12 +21,18 @@ class LlmService(pydantic.BaseModel):
...
+        raise NotImplementedError
+
     def get_chat_model(self, llm_model: str, temperature: float) -> BaseChatModel:
         raise NotImplementedError
 
     def transcribe_audio(self, audio: BytesIO) -> str:
         raise NotImplementedError
 
+    def get_callback_handler(self, llm_model: BaseLanguageModel) -> BaseCallbackHandler:
+        return TokenCountingCallbackHandler(llm_model)
```
Comment (proteusvacuum) [inline apps/service_providers/llm_service/main.py:34]: Curious if this should be a list? Maybe there's no need for that currently.

File: apps/service_providers/llm_service/main.py [hunk 3]
```diff
@@ -49,6 +63,11 @@ def transcribe_audio(self, audio: BytesIO) -> str:
         )
         return transcript.text
 
+    def get_callback_handler(self, llm_model: BaseLanguageModel) -> BaseCallbackHandler:
+        from langchain_community.callbacks import OpenAICallbackHandler
+
+        return OpenAICallbackHandler()
+
 
 class AzureLlmService(LlmService):
```
Comment (proteusvacuum) [inline apps/service_providers/llm_service/main.py:72]: Can this one use the OpenAICallbackHandler? It's hard to find in the docs...

**Review by proteusvacuum — COMMENTED** [review]
⛓️ 
Looking good!

File: apps/service_providers/llm_service/main.py (hunk 3 above)
Comment (snopoke) [reply apps/service_providers/llm_service/main.py:72]: I'm not sure. I don't think their API returns the token usage.

File: apps/service_providers/llm_service/main.py (hunk 2 above)
Comment (snopoke) [reply apps/service_providers/llm_service/main.py:34]: yea, it could be multiple

File: apps/chat/bots.py [hunk 4]
```diff
@@ -4,36 +4,24 @@
...
-                experiment_session=experiment_session,
-            )
-        else:
-            return BasicConversation(
-                prompt_str=prompt_str,
-                source_material=source_material,
-                memory=ConversationBufferMemory(return_messages=True),
-                llm=llm,
-            )
+        return BasicConversation(
```
Comment (SmittieC) [inline apps/chat/bots.py:19]: Since this will only be used for the safety boy now, I think the memory component of this object is redundant

**Review by SmittieC — APPROVED** [review]
This is super cool!

File: apps/chat/bots.py (hunk 4 above)
Comment (snopoke) [reply apps/chat/bots.py:19]: Yup, I didn't want to change too much in one PR.

**Pull request merged**